### PR TITLE
Fix BWC for file-settings based role mappings

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR modifies role mapping handling at the transport-layer to return to BWC behavior in ECK w.r.t. file-settings managed role mappings, after the changes in https://github.com/elastic/elasticsearch/pull/107410.

1. Include cluster-state role mappings (if any exist) in the response. Native role mappings will always be included first. Native role mappings and cluster-state role mappings may have duplicate names. 
2. Prevent Put Role Mapping calls for role mappings with a name matching a role mapping in cluster-state.
3. Include an error on  Delete Role Mapping calls if no native role mapping matching the name exist, but a cluster-state one does. 

Furthermore, the cluster-state role mapper is now enabled by default. If there are role mappings in cluster-state, these will apply during role resolution. This change should only affect ECK; cluster-state role mappings are only populated via file settings; these are only supported and documented in ECK and Serverless. In other environments, the cluster state role mapper will thus be empty and consulting it will be a noop.

Relates: ES-9628

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java:86]: These can definitely be cleaned up not to duplicate so much code but I ran out of time

File: x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecuritySettingsSource.java
Comment (n1v0lg) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecuritySettingsSource.java:406]: I think this should work but in case some test somewhere makes an explicit assumption this is `false` I'm fine with keeping it `false`

Comment (jfreden) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecuritySettingsSource.java:406]: If we remove this from serverless after this PR is merged we can come back and remove this property. Setting it to true here makes sense I think.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java:86]: Did some minor clean up. I think it's nice that `hasMapping` returns early.

**Comment by jfreden** [comment]
I've made some updates to the PR: 
- Check for cluster state blocks when reading cluster state mappings. The only blocking we can handle is global block on reading cluster state. There is no block on custom metadata as far as I can see. I think this handles the case where cluster state is not yet ready to be read (fully initilized?). Since we use `clusterService.state()` this could probably happen during startup. 
- Delete API should allow to delete native role mapping. The delete API doesn't fail even if you try to delete an operator configured role mapping. It just returns not found. I think that makes more sense than throwing an exception in this case. 
- The get role mappings API now return a `"metadata": {"_read_only": true}` flag if it comes from cluster state. The new field is added to the existing metadata of a role mapping only when returned in the get API. 
- General cleanup (minor, was already in a very good state)

TODO: 
- Add more unit/integration tests
- Do more manual testing with ECK 
- As a separate PR, create a security migration for removing role mappings with names that are both in cluster state and in the `.security` index as a one-time migration. This is to make sure that the roles that were accidentally moved to cluster state do not fall back to stale role mappings in the `.security` index if deleted.

**Comment by jfreden** [comment]
This only focuses on ECK since we’re not expecting any of this to impact Serverless (native role mappings are disabled there and we’re reading from cluster state to get operator defined role mappings). 

- Name clash - the same name for a role mapping is defined in both operator and native role mapping.
- Operator Defined - role mapping defined in settings.json (set through StackConfigPolicy)
- Native - role mapping in native role mapping store in the .security index 
- Effective Role Mapping - Role mapping being used during authorization.

### APIs
#### Get Role Mappings
Returns: Native + Operator Defined (with additional `metatdata _read_only: true`)

#### Delete Role Mapping
Deletes native role mapping (delete operator defined role mapping will result in 404)

#### Put Role Mapping
Write to .security index. Name clashes not prevented but warning header returned.

#### Effective Role Mapping
Operator Defined + Native (even if the same mapping name appears in both)

After this change the operator defined role mappings will exist both in the .security index and in cluster state for customers that use operator defined role mappings. The customers that made any changes to their operator defined role mappings in the .security index (native store) while the bug was active, will lose their changes. 

If an operator defined role mapping is removed, it will fall back to the stale one in the .security index, that’s why as part of this effort we’re also going to create a migration script that run once to clean up any operator defined role mappings that are present both in cluster state and in the .security index (this will be done by name).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java:47]: nit: why change from safeMap to delegateFailureAndWrap ? (is just style preference ?)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java:60]: (can be done in a followup) I think this is always enabled, so we can remove this setting entirely. We will need to (in order) a) remove from serverless config b) remove from serverless code c) remove here

alternatively, we can register the setting in main code and allow users to disable it, but will also need to test and throw exceptions if the setting is false and K8's defined the configuration (to avoid a write, but never read situation).

I would prefer to just remove it entirely since there isn't a good reason to disable it.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java
Comment (jakelandis) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:318]: is this testing that the cluster state role mappings are added then deleted as expected ? if so, can you rename this to `testClusterStateRoleMappingsAddedThenDeleted` (or the like)

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java
Comment (jakelandis) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:373]: nit: can you assert that ones from cluster state are read only and the other ones are not is not. Assuming we accept the in flight discussion to append `(read only)` to the name, the order is not required.

**Review by jakelandis — COMMENTED** [review]
Looking good, all minor comments

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java]: I changed this back to `safeMap`. Not sure about the intention here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java]: Change this to `(read only)` suffix.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportPutRoleMappingAction.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportPutRoleMappingAction.java:60]: nit: use `private static final` class level variable to avoid the (tiny) allocation for new string everytime (but moreso for style conventions and DRY)

Also, super nitpick: I don't think we need to enforce the white space in the check... so we should only check endsWith `"(read only)"` not `" (read only)"` (my only reason is that it begs the question...is there some meaning to the white space ? (which the answer is no)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:76]: nit: use a static constant (preferably the same one used in by put)

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java
Comment (jakelandis) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:391]: nit: can you add this final assertion : 
```
 var request = new GetRoleMappingsRequest();
        var response = client().execute(GetRoleMappingsAction.INSTANCE, request).get();
        assertTrue(response.hasMappings());
        assertThat(
            Arrays.stream(response.mappings()).map(ExpressionRoleMapping::getName).toList(),
            containsInAnyOrder(
                "_everyone_kibana",   
                "zzz_mapping",
                "123_mapping"
            )
        );
 ```     
(I think that is what is left ? I havent tested)

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java
Comment (jakelandis) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:390]: > INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING

? is that the correct key ?

**Review by jakelandis — APPROVED** [review]
LGTM (pending CI)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportPutRoleMappingAction.java
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportPutRoleMappingAction.java]: Yes, sorry about this one. Moved this to a constant.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java
Comment (jfreden) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:391]: It's part of the test data and validated in other tests. The `testJSON` has this setting but the `emptyJSON` doesn't, so it should not be there after the `emptyJSON` has been loaded. It's a little distracting, but probably good to keep the validation.

**Comment by jfreden** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/114197)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Comment by jfreden** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/114224)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**
